v6/v6_interface.ml

```diff
@@ -12,33 +12,139 @@
  * GNU Lesser General Public License for more details.
  *)
 
+(**
+   @group Licensing
+*)
+open Rpc
+open Idl
+
 let service_name = "v6d"
-let queue_name = ref (Xcp_service.common_prefix ^ service_name)
+let queue_name = (Xcp_service.common_prefix ^ service_name)
 let default_sockets_dir = "/var/lib/xcp"
 let default_path = ref (Filename.concat default_sockets_dir service_name)
 let uri () = "file:" ^ !default_path
+let json_path = "/var/xapi/v6.json"
+let xml_path = "/var/xapi/v6"
+
+(* --- global types used in most API calls --- *)
 
+(** Uninterpreted/sanitised string associated with operation *)
 type debug_info = string
+[@@deriving rpcty]
+
+(** Record representing software edition *)
+type edition =
+  {
+    ed_title : string ;
+    (** Name of edition, this will be passed to apply_edition *)
+    ed_official_title : string ;
+    (** Marketing title used to advertise edition *)
+    ed_code : string ;
+    (** Abbreviated form of name, used to show up in logs and on command line *)
+    ed_order : int ;
+    (** Number indicating ordering among other editions
+        with low numbers corresponding to editions with fewer features, and vice versa *)
+  }
+[@@deriving rpcty]
+
+(** List of edition records *)
+type edition_list = edition list
+[@@deriving rpcty]
 
-type edition_info = {
-    edition: string;
+(** Record of edition info, including xapi parameters and features *)
+type edition_info =
+  {
+    edition_name: string;
+    (** Name of edition *)
     xapi_params: (string * string) list;
+    (** List of parameters used by Xapi *)
     additional_params: (string * string) list;
+    (** Addition parameters supplied *)
     experimental_features: (string * bool) list;
-}
+    (** List of experimental features and whether they're available in this edition *)
+  }
+[@@deriving rpcty]
+
+(** [string * string] list *)
+type stringPairLst = (string * string) list
+[@@deriving rpcty]
+
+(* --- errors wrapped in generic errors type --- *)
+
+(** Wrapper for specific licensing errors *)
+type errors =
+  | Invalid_edition of string
+  (** Thrown by apply_edition on receiving unfamiliar edition. Note: get_editions returns list of all valid editions *)
+  | License_expired
+  (** Thrown by license_check when expiry date matches or precedes current date *)
+  | License_processing_error
+  (** License could not be processed *)
+  | License_checkout_error of string
+  (** License could not be checked out *)
+  | Missing_connection_details
+  (** Thrown if connection port or address parameter not supplied to check_license *)
+  | V6d_failure
+  (** Licensing daemon failed *)
+[@@default V6d_failure]
+[@@deriving rpcty]
+(* --- API interface --- *)
+
+(** Generic licensing error of error variant type *)
+exception LicensingError of errors
+[@@deriving rpcty]
+
+(** handle exception generation and raising *)
+let err = Error.{
```

Comment:
A PR to `ocaml-rpc` to autogenerate `err` passing the exception wrapper type into a functor is welcome :)